Update ray tracing sample to use VK_KHR_ray_tracing

SaschaWillems opened the pull request:
## Description

This PR updates the basic ray tracing sample to use the newly introduced cross-vendor VK_KHR_ray_tracing extension (instead of the vendor-specific VK_NV_ray_tracing).

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build and run on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- n/a If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- n/a Any dependent assets have been merged and published in downstream modules

## Notes
- Right now only NVIDIA supports the new extension, so I only tested it on a current NVIDIA GPU (RTX 2060) with current drivers
- The sample requires new functionality in the volk and VMA sub moduls, so I had to update these
- The setup for VMA has been extended with support for buffer device addresses (see the device class)
- The extension is currently beta, which requires ```VK_ENABLE_BETA_EXTENSIONS``` to be enabled. I added this to the CMAKE setup and enabled it by default

SaschaWillems committed:
Add support for validation with VK_EXT_Debug_Utils

Will fall back to the old VK_EXT_Debug_Report extensions if not supported

SaschaWillems committed:
Add new extensions sample for VK_EXT_Debug_Utils

SaschaWillems committed:
Destroy debug utils callback

SaschaWillems committed:
Only use debug utils functions of extension is available

SaschaWillems committed:
Add name to skysphere texture image

Add queue label

SaschaWillems committed:
Added tutorial for using the debug utilities extension with a debugger like RenderDoc

SaschaWillems committed:
Fix clang format and copyright headers

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into debug_utils

SaschaWillems committed:
Check against actually enabled instance extension to determine if debug utils can be used

Display a clear warning with reasons if debug utilities are not available

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into debug_utils

# Conflicts:
#	README.md
#	framework/core/instance.cpp

SaschaWillems committed:
Updated debug utils creation to use pNext instance structure chaining

SaschaWillems committed:
Updated readme

Fix spelling, add links

SaschaWillems committed:
Subpass dependencies

SaschaWillems committed:
Clang format

SaschaWillems committed:
Added hints to image markings

SaschaWillems committed:
Expanded setup chapter with differences between implicit and explicit debug utilities messenger creation

SaschaWillems committed:
Explicitly create callback, update readme

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into debug_utils

# Conflicts:
#	README.md
#	framework/core/instance.cpp

SaschaWillems committed:
Add debug utilities sample to api readme

SaschaWillems committed:
Use add_sample for CMake

SaschaWillems committed:
Fixed typo

SaschaWillems committed:
Properly initialize sub pass dependencies

SaschaWillems committed:
Adjusted subpass dependency access masks

SaschaWillems committed:
Porting VK_NV_ray_tracing to VK_KHR_ray_tracing

Updates submodules to support VK_KHR_ray_tracing
work-in-progress

SaschaWillems committed:
Porting VK_NV_ray_tracing to VK_KHR_ray_tracing

Added required features and vma settings to device creation
work-in-progress

SaschaWillems committed:
CMake flag for enabling Vulkan beta features

SaschaWillems committed:
Porting VK_NV_ray_tracing to VK_KHR_ray_tracing

Updated shaders
work-in-progress

SaschaWillems committed:
Porting VK_NV_ray_tracing to VK_KHR_ray_tracing

Updated sample code
work-in-progress

SaschaWillems committed:
Porting VK_NV_ray_tracing to VK_KHR_ray_tracing

Primitive count for top level AS
work-in-progress

SaschaWillems committed:
Porting VK_NV_ray_tracing to VK_KHR_ray_tracing

Fixed wrong geometry instance data sourcing
work-in-progress

SaschaWillems committed:
Enable Vulkan beta features for CMake

SaschaWillems committed:
First working version with VK_KHR_ray_tracing

Code restructuring

SaschaWillems committed:
Delete scratch buffers

Code cleanup

SaschaWillems committed:
Enable raytracing at device level

SaschaWillems committed:
Code cleanup, NV_RT extension no longer required for validation

SaschaWillems committed:
Code cleanup, restructuring and comments

SaschaWillems committed:
Merge branch 'master' into VK_KHR_ray_tracing

# Conflicts:
#	framework/core/device.cpp
#	framework/core/instance.cpp
#	samples/extensions/debug_utils/debug_utils.cpp
#	samples/extensions/debug_utils/debug_utils.h
#	samples/extensions/raytracing_basic/raytracing_basic.cpp

SaschaWillems committed:
Revert changes to device class required to get the VK_KHR_ray_tracing sample running

SaschaWillems committed:
Updated VK_KHR_ray_tracing sample to use new way of passing extension features to device creation

SaschaWillems committed:
Set VMA buffer address allocator flag if VK_KHR_Buffer_Device_Address extension has been requested

SaschaWillems committed:
Properly handle resize

SaschaWillems committed:
Copyright headers

eloj commented on samples/extensions/raytracing_basic/raytracing_basic.cpp:
```diff
@@ -100,279 +200,260 @@ void RaytracingBasic::create_storage_image()
 
 /*
-	Create the Shader Binding Table that binds the programs and top-level acceleration structure
+	Create the Shader Binding Table that holds the ray tracing piepeline's shader group handle
```
Comment typo.

SaschaWillems committed:
Merge remote-tracking branch 'upstream/master' into VK_KHR_ray_tracing

gary-sweet reviewed (changes requested):
I see some build warnings:

```
In file included from /build/gsweet/builds/7216_64/libs/vulkan/third_party_demos/KhronosVulkanSamplesExt/src/KhronosVulkanSamples_internal-build/app/samples.h:37:0,
                 from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/app/vulkan_samples.h:23,
                 from /build/gsweet/builds/7216_64/libs/vulkan/third_party_demos/KhronosVulkanSamplesExt/src/KhronosVulkanSamples_internal-build/app/main.cpp:23:
/build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/app/../samples/extensions/raytracing_basic/raytracing_basic.h:24:0: warning: "VK_KHR_ray_tracing" redefined
 #define VK_KHR_ray_tracing
 
In file included from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/third_party/vulkan/include/vulkan/vulkan.h:88:0,
                 from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/third_party/volk/volk.h:50,
                 from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/framework/common/strings.h:23,
                 from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/framework/common/error.h:24,
                 from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/framework/common/utils.h:20,
                 from /build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/framework/platform/platform.h:24,
                 from /build/gsweet/builds/7216_64/libs/vulkan/third_party_demos/KhronosVulkanSamplesExt/src/KhronosVulkanSamples_internal-build/app/main.cpp:21:
/build/gsweet/vc5_git/thirdparty_vulkan/Vulkan-Samples/third_party/vulkan/include/vulkan/vulkan_beta.h:85:0: note: this is the location of the previous definition
 #define VK_KHR_ray_tracing 1
```

SaschaWillems commented:
Thanks for the hint. This was a left-over trying to get volk include the beta functions for ray tracing. Will try to fix this asap.

SaschaWillems committed:
Fixed typo

SaschaWillems committed:
Removed duplicate define

VK_KHR_ray_tracing is already defined in the beta headers when included

SaschaWillems committed:
Always enable VK_ENABLE_BETA_EXTENSIONS instead of having this depend on a CMake option

SaschaWillems committed:
Explicit type cast

Fixes compiler warning

SaschaWillems commented:
I have pushed a few updates. The warning reported by  @gary-sweet is now fixed along with another compiler warning due to an implicit type-cast in the sample.

I also removed the VKB CMake option to enable Vulkan beta extension and always enabled it, as disabling would break the build.

aurolou reviewed (changes requested):
Could you please add a warning that this extension is still in beta and will be subject to change? Thanks

SaschaWillems commented:
> Could you please add a warning that this extension is still in beta and will be subject to change? Thanks

Would a note at the top of the sample's source file be sufficient?

ghost commented:
> > Could you please add a warning that this extension is still in beta and will be subject to change? Thanks
> 
> Would a note at the top of the sample's source file be sufficient?

Yes, this should be enough in my opinion, please make the update.

aurolou commented:
> > > Could you please add a warning that this extension is still in beta and will be subject to change? Thanks
> > 
> > 
> > Would a note at the top of the sample's source file be sufficient?
> 
> Yes, this should be enough in my opinion, please make the update.

yes, should be enough. we could also mention that development drivers are needed. it's just to make sure people are not trying to ship ray tracing code as production :)

SaschaWillems committed:
Add note about beta state of the extension

SaschaWillems commented:
I have added the following comment to the header and implementation file:

```cpp
/*
 * Important note:
 *
 * The VK_KHR_ray_tracing extension is currently considered a BETA extension.
 *
 * This means that it is not yet production ready and subject to change until it's finalized.
 *
 * In order to use this sample you may also need special developer drivers.
 */
```

Will remove it once the extension is finalized.

I hope this makes it clear that the extension is not yet ready for production.
